### PR TITLE
CSDS: add new config status ACKED

### PR DESCRIPTION
**Pull request by lidizheng**
Signed-off-by: Lidi Zheng <lidiz@google.com>

Signed-off-by: Lidi Zheng <lidiz@google.com>

Commit Message: Add new config status for CSDS: ACKED.
Additional Description: As the CSDS service definition described, it has the potential to be used to expose xDS config from a client or proxy. gRPC wants to utilize this service to improve its debuggability. But the `ConfigStatus` is designed from the control plane point of view. Especially, the client cannot predict if there is new config on its way, so it can't accurately claim any xDS config status as SYNCED. We need another config status to indicate the status that the client received the status and sent out ACK. The actual wording is up to debate.
Risk Level: Low
Testing: Not needed
Docs Changes: Inline with proto definition
Release Notes: New config status for CSDS: ACKED.

CC @fuqianggao @markdroth @menghanl

### CODE REVIEW

File: api/envoy/service/status/v2/csds.proto
Comment (htuch) [inline api/envoy/service/status/v2/csds.proto:57]: v2 is frozen, so requires justification for change here. Do you need this before we move to v3?

File: api/envoy/service/status/v3/csds.proto
Comment (htuch) [inline api/envoy/service/status/v3/csds.proto:57]: I'm almost thinking we want separate status enum messages or very explicit field naming to help make clear which one (client vs. server status) we're talking about.

File: api/envoy/service/status/v2/csds.proto
Comment (markdroth) [reply api/envoy/service/status/v2/csds.proto:57]: I don't think we need this in v2.  Let's do it for v3 only.

File: api/envoy/service/status/v3/csds.proto
Comment (markdroth) [inline api/envoy/service/status/v3/csds.proto:45]: s/haven't/hasn't/

File: api/envoy/service/status/v3/csds.proto
Comment (markdroth) [reply api/envoy/service/status/v3/csds.proto:57]: Yeah, seeing these comments, I guess I do think that having a separate set of enums would be clearer.  We can call them something like `CLIENT_REQUESTED`, `CLIENT_ACKED`, and `CLIENT_NACKED`.

We can either add these as a new values in this existing enum, or we can add a whole separate enum field for use in clients.

**Review by markdroth — COMMENTED** [review]
Thanks for doing this, Lidi!

**Comment by htuch** [comment]
Let's split this out as a separate enum. I feel it's going to be necessary for any CSDS client to have some logic to make sense of whether they are interpreting a management or client status here.

File: api/envoy/service/status/v2/csds.proto
Comment (lidizheng) [reply api/envoy/service/status/v2/csds.proto]: Changed in v2 reverted.

File: api/envoy/service/status/v3/csds.proto
Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto]: Adding a new proto enum struct named ClientSideStatusConfig. It's a bit verbose, but I think it needs to have a very clear naming.

**Review by lidizheng — COMMENTED** [review]
@markdroth PTALA.

File: api/envoy/service/status/v3/csds.proto
Comment (markdroth) [inline api/envoy/service/status/v3/csds.proto:68]: Let's explicitly document that the resource content that is returned in this case will be the most recent version accepted, if any, not the version that was NACKed.

File: api/envoy/service/status/v3/csds.proto
Comment (markdroth) [inline api/envoy/service/status/v3/csds.proto:91]: It's probably worth adding a comment here that `status` will be populated by xDS servers but `client_side_config_status` will be populated by xDS clients.

File: api/envoy/service/status/v3/csds.proto
Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto]: Documented the expected config dump behavior here and in the PerXdsConfig message.

File: api/envoy/service/status/v3/csds.proto
Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto:94]: Comments added.

**Review by lidizheng — COMMENTED** [review]
@markdroth  PTALAA.

File: api/envoy/service/status/v3/csds.proto
Comment (markdroth) [inline api/envoy/service/status/v3/csds.proto:69]: s/NACK/NACKed/

File: api/envoy/service/status/v3/csds.proto
Comment (markdroth) [inline api/envoy/service/status/v3/csds.proto:70]: s/should remain/will be/

File: api/envoy/service/status/v3/csds.proto
Comment (markdroth) [inline api/envoy/service/status/v3/csds.proto:93]: s/is//

Comment (markdroth) [inline api/envoy/service/status/v3/csds.proto:93]: Please add "Will not be present if the CSDS server is an xDS client."

File: api/envoy/service/status/v3/csds.proto
Comment (markdroth) [inline api/envoy/service/status/v3/csds.proto:96]: Instead of "Only one of the config status should present in one PerXdsConfig message", let's just say "Will not be present if the CSDS server is an xDS server."

File: api/envoy/service/status/v3/csds.proto
Comment (markdroth) [inline api/envoy/service/status/v3/csds.proto:100]: Suggest just calling this `client_status`.

File: api/envoy/service/status/v3/csds.proto
Comment (markdroth) [inline api/envoy/service/status/v3/csds.proto:57]: Suggest calling this `ClientConfigStatus`.

File: api/envoy/service/status/v3/csds.proto
Comment (markdroth) [reply api/envoy/service/status/v3/csds.proto:94]: (It's too bad that we can't use a `oneof` here, but that would be a breaking change for the go proto API.)

**Review by markdroth — COMMENTED** [review]
Please also change the comment on line 23-26 to indicate that we are now going to use this on the xDS client side.

Thanks!

File: api/envoy/service/status/v3/csds.proto
Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto]: Changed to `ClientConfigStatus`.

In this proto definition, there is the `ClientConfig` message. It might be a bit confusing to users that why `ClientConfigStatus` is not the status for the `ClientConfig` message, but the client-side view of config status. On the other hand, the `ClientSideConfigStatus` is not much better.

Better naming is still welcomed. Picking `ClientConfigStatus` since it is simpler.

File: api/envoy/service/status/v3/csds.proto
Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto]: Good catch!

File: api/envoy/service/status/v3/csds.proto
Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto]: Replaced.

File: api/envoy/service/status/v3/csds.proto
Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto]: Deleted.

Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto]: Comment added.

File: api/envoy/service/status/v3/csds.proto
Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto]: Done. That's better.

File: api/envoy/service/status/v3/csds.proto
Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto]: Name changed.

**Review by lidizheng — COMMENTED** [review]
@markdroth Thanks for the quick review! PTALAAA.

**Comment by markdroth** [comment]
Please also change the comment on line 23-26 to indicate that we are now going to use this on the xDS client side.  Thanks!

**Review by markdroth — APPROVED** [review]
This looks great!

I'll recuse myself and let @htuch or one of the other API shepherds give API approval.

**Comment by markdroth** [comment]
(Whoops -- I didn't realize that my approving the PR would satisfy the API approval requirement; I thought I had to say "/lgtm api".  But since one of the maintainers has to review before merging anyway, that's probably okay.)

File: api/envoy/service/status/v3/csds.proto
Comment (htuch) [inline api/envoy/service/status/v3/csds.proto:69]: Can we have a similar statement added about control plane behavior for configs with a negative status, i.e. "this is always the latest successful config?" Or maybe that's not even desired, we might want the latest successful and unsuccessful config? Would that also make sense for client? CC @fuqianggao

**Review by htuch — COMMENTED** [review]
Needs a fix_format.

File: api/envoy/service/status/v3/csds.proto
Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto:69]: This behavior wasn't clearly described in this proto definition before, or in other documentation. Could it be a control-plane specific detail?

Hi @fuqianggao, can you help to explain the designed behavior of config dump when client replied with NACK?

File: api/envoy/service/status/v3/csds.proto
Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto]: Annotation added, TIL.

Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto]: @htuch Can you advice what should we use in v4alpha? The format suggests that the migration annotation is expected to be expanded into real `oneof` in v4alpha. But that will break backward compatibility.

Comment (markdroth) [reply api/envoy/service/status/v3/csds.proto]: That's fine -- v4alpha will be a new major version (if it ever actually happens), so breaking changes are okay.  This is actually the whole point of the annotation, to make sure this change happens when the next major version bump occurs.

Comment (markdroth) [reply api/envoy/service/status/v3/csds.proto]: Oh, but it looks like this didn't actually work for some reason.  The annotation was carried forward to v4alpha instead of being turned into a oneof.  Maybe the annotation has the wrong syntax or something?

File: api/envoy/service/status/v3/csds.proto
Comment (htuch) [inline api/envoy/service/status/v3/csds.proto:100]: I think you might need to have a `oneof` existing. I suggest moving this field lexically before the existing `status` and wrapping it in a `oneof status_config`. That should allow the v4alpha generation to work.

**Review by htuch — COMMENTED** [review]
LGTM modulo the thread on clarifying management server behavior on stale/nacked config. I'd like there to be consistency, and ideally we should have a plan for how to support reporting both last known good and bad config (at least on management server side), since both are useful, CC @fuqianggao

File: api/envoy/service/status/v3/csds.proto
Comment (fuqianggao) [reply api/envoy/service/status/v3/csds.proto:70]: For control plane, it will be the latest config that control plane has. If status is NACK, it means the config is sent to client, but rejected.

Comment (htuch) [reply api/envoy/service/status/v3/csds.proto:70]: Makes sense. I think the actionable thing is for @lidizheng to update the management server wording to reflect this. I'm still wondering in this discussion thread, is there value in having both active and failed configs reported?

Comment (markdroth) [reply api/envoy/service/status/v3/csds.proto:70]: I suspect that for both xDS servers and xDS clients, there would be additional overhead in storing old versions of configs.  In general, both sides will store only the latest version that they care about, so they don't have anything else to report here.

Comment (menghanl) [reply api/envoy/service/status/v3/csds.proto:70]: And if all clients implement CSDS, the CLI can get the active config from the clients, and the most up-to-date config from the management server (if they are different).

Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto:72]: I have updated the comment for `ConfigStatus.ERROR` to note that the management server will dump the latest config in case of client replying NACK.

**Comment by lidizheng** [comment]
@htuch PTALAA. All tests are passing, and the mentioned behavior is documented.

File: api/envoy/service/status/v3/csds.proto
Comment (markdroth) [reply api/envoy/service/status/v3/csds.proto:102]: Lidi, I think you still need to address this comment.

Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto:102]: According to the style in the code base, the `oneof_promotion ` annotation seems appear in pairs ([example 1](https://github.com/envoyproxy/envoy/blob/99471fd8f4ef7406f50cf41cf1cfa22bbdfeacc7/api/envoy/extensions/transport_sockets/tls/v3/secret.proto#L36), [example 2](https://github.com/envoyproxy/envoy/blob/99471fd8f4ef7406f50cf41cf1cfa22bbdfeacc7/api/envoy/extensions/common/tap/v3/common.proto#L36)). Splitting the oneof options into different layers might breaks the consistency, and slightly weaken the semantic intention of the upcoming migration.

So, in my mind, keeping `oneof_promotion` in v3 and real oneof in v4alpha is the better way.

Comment (markdroth) [reply api/envoy/service/status/v3/csds.proto:102]: Yes, that's what I meant.

The v4alpha file looks right now.  The last time I looked, I thought it was still wrong, but maybe I was looking at an old snapshot.

Anyway, this looks great.  Thanks!

File: api/envoy/service/status/v3/csds.proto
Comment (htuch) [reply api/envoy/service/status/v3/csds.proto:72]: I like the idea of being able to query both client and server CSDS and effectively determine the failing and active config. @lidizheng can you please file an issue for Envoy to add support for CSDS? You don't have to do the implementation, but we should track this as an open feature request that we can share with the community.

**Review by htuch — APPROVED** [review]

File: api/envoy/service/status/v3/csds.proto
Comment (lidizheng) [reply api/envoy/service/status/v3/csds.proto:72]: Filed as an issue: https://github.com/envoyproxy/envoy/issues/13181. For Envoy, it overlaps with the admin config dump feature. Today, tools like `istioctl` is fetching xDS config from server via CSDS, and from client via admin config dump (HTTP).

**Pull request merged**
